staging/src/k8s.io/kms/encryption/service_test.go

```diff
@@ -0,0 +1,364 @@
+				Annotations: map[string][]byte{
+					"version.encryption.remote.io":     []byte("1"),
+					"key-version.encryption.remote.io": []byte("2"),
+				},
+			},
+			want: nil,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			got := validateRemoteKMSResponse(tc.input)
+			if got != tc.want {
+				t.Errorf("validateRemoteKMSResponse(%v) = %v, want %v", tc.input, got, tc.want)
+			}
+		})
+	}
+}
+
+var _ service.Service = &testRemoteService{}
+
+type testRemoteService struct {
+	mu sync.Mutex
+
+	keyID    string
+	disabled bool
+}
+
+func (s *testRemoteService) Encrypt(ctx context.Context, uid string, plaintext []byte) (*service.EncryptResponse, error) {
+	if s.disabled {
+		return nil, errors.New("failed to encrypt")
+	}
+	return &service.EncryptResponse{
+		KeyID:      s.keyID,
+		Ciphertext: []byte(base64.StdEncoding.EncodeToString(plaintext)),
+		Annotations: map[string][]byte{
+			"version.encryption.remote.io": []byte("1"),
+		},
+	}, nil
+}
+
+func (s *testRemoteService) Decrypt(ctx context.Context, uid string, req *service.DecryptRequest) ([]byte, error) {
+	if s.disabled {
+		return nil, errors.New("failed to decrypt")
+	}
+	if len(req.Annotations) != 1 {
+		return nil, errors.New("invalid annotations")
+	}
+	return base64.StdEncoding.DecodeString(string(req.Ciphertext))
+}
+
+func (s *testRemoteService) Status(ctx context.Context) (*service.StatusResponse, error) {
+	if s.disabled {
+		return nil, errors.New("failed to get status")
+	}
+	return &service.StatusResponse{
+		Version: "v2alpha1",
+		Healthz: "ok",
+		KeyID:   s.keyID,
+	}, nil
+}
+
+func (s *testRemoteService) SetDisabledStatus(disabled bool) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.disabled = true
+}
+
+func TestEncrypt(t *testing.T) {
+	remoteKMS := &testRemoteService{keyID: "test-key-id"}
+	localKEKService := NewLocalKEKService(remoteKMS)
+
+	validateResponse := func(got *service.EncryptResponse, t *testing.T) {
+		if len(got.Annotations) != 2 {
+			t.Fatalf("Encrypt() annotations = %v, want 2 annotations", got.Annotations)
+		}
+		if _, ok := got.Annotations[referenceKEKAnnotationKey]; !ok {
+			t.Fatalf("Encrypt() annotations = %v, want %v", got.Annotations, referenceKEKAnnotationKey)
+		}
+		if got.KeyID != remoteKMS.keyID {
+			t.Fatalf("Encrypt() keyID = %v, want %v", got.KeyID, remoteKMS.keyID)
+		}
+		if localKEKService.localTransformer == nil {
+			t.Fatalf("Encrypt() localTransformer = %v, want non-nil", localKEKService.localTransformer)
+		}
+	}
+
+	// local KEK is generated and encryption is successful
+	got, err := localKEKService.Encrypt(context.Background(), "test-uid", []byte("test-plaintext"))
+	if err != nil {
+		t.Fatalf("Encrypt() error = %v", err)
+	}
+	validateResponse(got, t)
+
+	// local KEK is used for encryption even when remote KMS is failing
+	remoteKMS.SetDisabledStatus(true)
+	if got, err = localKEKService.Encrypt(context.Background(), "test-uid", []byte("test-plaintext")); err != nil {
```

Comment:
This applies to call contexts - avoid using the background context, use `context.TODO()` or a context that is canceled at the end of the test (with some timeout).